File: scala-solution/Soldier.scala
import cwinter.codecraft.core.api._
import cwinter.codecraft.util.maths.Vector2
import scala.util.Random


class Soldier extends DroneController {
  override def onTick(): Unit = {
    val enemies = dronesInSight.filter(_.isEnemy)
    enemies.headOption match {
      case Some(enemy) =>
        moveTo(enemy.position)
        if (missileCooldown <= 0 && isInMissileRange(enemy))
          fireMissilesAt(enemy)
      case None =>
        if (!isMoving) {
          val randomDirection = Vector2(2 * math.Pi * Random.nextDouble())
          val targetPosition = position + 1000 * randomDirection
          moveTo(targetPosition)
        }
    }
  }
}

File: java-solution/Soldier.java
import cwinter.codecraft.core.api.*;
import cwinter.codecraft.util.maths.Vector2;
import java.util.Random;


class Soldier extends JDroneController {
  static Random rng = new Random();


  @Override public void onTick() {
    for (Drone drone : dronesInSight()) {
      if (drone.isEnemy()) {
        moveTo(drone.position());
        if (missileCooldown() == 0 && isInMissileRange(drone)) {
          fireMissilesAt(drone);
        }
        return;
      }
    }

    if (!isMoving()) {
      Vector2 randomDirection = new Vector2(2 * Math.PI * rng.nextDouble());
      Vector2 targetPosition = position().plus(randomDirection.times(500));
      moveTo(targetPosition);
    }
  }
}

